lib/OptimizationMOI/src/moi.jl

```diff
@@ -0,0 +1,492 @@
+mutable struct MOIOptimizationCache{F <: OptimizationFunction, uType, P, LB, UB, I, S, EX,
+                                    CEX, O} <: SciMLBase.AbstractOptimizationCache
+    f::F
+    u0::uType
+    p::P
+    lb::LB
+    ub::UB
+    int::I
+    sense::S
+    expr::EX
+    cons_expr::CEX
+    opt::O
+    solver_args::NamedTuple
+end
+
+function MOIOptimizationCache(prob::OptimizationProblem, opt; kwargs...)
+    isnothing(prob.f.sys) &&
+        throw(ArgumentError("Expected an `OptimizationProblem` that was setup via an `OptimizationSystem`, consider `modelingtoolkitize(prob).`"))
+
+    # TODO: check if the problem is at most bilinear, i.e. affine and or quadratic terms in two variables
+    expr_map = get_expr_map(prob.f.sys)
+    expr = repl_getindex!(convert_to_expr(MTK.subs_constants(MTK.objective(prob.f.sys)),
+                                          prob.f.sys; expand_expr = false, expr_map))
+
+    cons = MTK.constraints(prob.f.sys)
+    cons_expr = Vector{Expr}(undef, length(cons))
+    Threads.@sync for i in eachindex(cons)
+        Threads.@spawn cons_expr[i] = repl_getindex!(convert_to_expr(Symbolics.canonical_form(MTK.subs_constants(cons[i])),
+                                                                     prob.f.sys;
+                                                                     expand_expr = false,
+                                                                     expr_map))
```

Comment:
MTK is needed here

Reply:
Although both are in the `prob.expr` and `prob.cons_expr`, the parameters are replaced in MTK:
https://github.com/SciML/ModelingToolkit.jl/blob/master/src/systems/optimization/optimizationsystem.jl#L497-L498